src/analysis/souffle/examples/multimic/multimic-variant.arcs

```diff
@@ -0,0 +1,69 @@
+// To be able to handle this use-case, since we don't have information about 
+// policies or who is even in the room until runtime, we need to evaluate these 
+// checks at runtime. The taint analysis could work similarly to the static one 
+// we have now, but it happens at runtime: reading from handles causes the tags 
+// on that handle to propagate to the write handles. Downgrades still remove 
+// tags.
+
+// Represents an actual microphone within a smart mic listening to audio.
+particle MicIn
+  audio: writes AudioData { }
+  // MicIn says hasTag("audio", unidentified_people::userSpeech)
+  // NOTE: at runtime, this also receives facts that add tags for the 
+  // identified users from "MobileDeviceRuntime" (see also comment in 
+  // multimic.authlogic).
+
+// Storage credentials can be in the form of authorization logic facts
+// and guarded by the may/will check. So I didn't include credentials checking 
+// here.
+
+particle Notifier
+  audioIn: reads AudioData { }
+  networkChannelsToUserDevices: writes Notification { }
+  // (This could be more like "claims { ... } " to keep it more in Arcs style.)
+  // Notifier says {
+  //    will("deliver_privacy_notifications", "audioIn").
+  //
+  // }
+
+// Represents the CPU within the mic.
+particle MicCompute
+  audioIn: reads AudioData { }
+  asrNeuralNet: reads ASRNeuralNet { }
+  asrStorageConn: writes ASRData { }
+  // Need to evaluate this at runtime since information about user's
+  // policies are not known until then.
+  // MicCompute says {
+  //    "MobileDeviceRuntime" canSay isInRoom(UserX)
+  //        :- isUser(UserX). // we would need to figure out how to get rid of 
+  //                          // this isUser part
```

Comment:
One thought that I've had about Universes: are Universes in our policy language the same thing as types? Not user-defined types, but the basic types that our policy language understands.

One way to look at a type is that it is an overapproximation of the set of possible values that may be associated with a particular name. A Universe in the Datalog sense is just the exact set of data values that a particular family of names used in recipes can have. I don't think it's a coincidence that we have almost a one-to-one correspondence between Datalog base types and Datalog universes. Perhaps in the policy language that we are building the two concepts should be unified.

Reply:
I'm not totally sure I completely follow your point. I think the universe relation pattern we have is just our way of making sure our code is stratifiable. I do think that once we have the key system design details figured out, having types in our policy language will be useful. I remember you also had some ideas about subtypes for principals, and I think that still sounds potentially useful.

Actually, is this your point:
 - programmer-introduced types tend to correspond to the universe relations that we make
 - if we want to eventually generate and populate universe relations "behind-the-scenes" so that our developer-facing language is clearer and less verbose, then the user-defined types are probably exactly the universe relations we need to create. (And the constants that get used for each user-defined type are probably how we populate the universe relations).
 
 If that's close to your point, then I think that's on to something.